pkg/engine/lifecycletest/target_test.go

```diff
@@ -89,6 +89,418 @@ func TestDestroyTarget(t *testing.T) {
+	})
+
+	snap, err := lt.TestOp(Update).RunStep(project, p.GetTarget(t, nil), opts, false, p.BackendClient, nil, "0")
+	require.NoError(t, err)
+
+	require.Len(t, snap.Resources, 2)
+	assert.Equal(t, snap.Resources[0].URN.Name(), "default") // provider
+	assert.Equal(t, snap.Resources[1].URN.Name(), "resA")
+}
+
+// We should be able to create a simple `A > B > C` hierarchy and destroy
+// `C` and by excluding `A` and `B`. We should then be able to exclude the
+// provider and destroy everything under it.
+func TestDestroyExcludeTarget(t *testing.T) {
+	t.Parallel()
+
+	p := &lt.TestPlan{}
+	project := p.GetProject()
+
+	// Operation 1: deploy the three-item hierarchy
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+	}
+
+	program := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		resA, err := monitor.RegisterResource("pkgA:m:typA", "resA", true)
+		assert.NoError(t, err)
+
+		resB, err := monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{Parent: resA.URN})
+		assert.NoError(t, err)
+
+		_, err = monitor.RegisterResource("pkgA:m:typA", "resC", true, deploytest.ResourceOptions{Parent: resB.URN})
+		assert.NoError(t, err)
+
+		return nil
+	})
+
+	hostF := deploytest.NewPluginHostF(nil, nil, program, loaders...)
+	opts := lt.TestUpdateOptions{T: t, HostF: hostF}
+
+	snap, err := lt.TestOp(Update).RunStep(project, p.GetTarget(t, nil), opts, false, p.BackendClient, nil, "0")
+	require.NoError(t, err)
+
+	require.Len(t, snap.Resources, 4)
+	assert.Equal(t, snap.Resources[0].URN.Name(), "default") // provider
+	assert.Equal(t, snap.Resources[1].URN.Name(), "resA")
+	assert.Equal(t, snap.Resources[2].URN.Name(), "resB")
+
+	opts.UpdateOptions.Excludes = deploy.NewUrnTargetsFromUrns([]resource.URN{
+		"urn:pulumi:test::test::pkgA:m:typA::resA",
+		"urn:pulumi:test::test::pkgA:m:typA$pkgA:m:typA::resB",
+	})
+
+	snap, err = lt.TestOp(Destroy).RunStep(project, p.GetTarget(t, snap), opts, false, p.BackendClient, nil, "1")
+
+	require.NoError(t, err)
+	require.Len(t, snap.Resources, 3)
+	assert.Equal(t, snap.Resources[0].URN.Name(), "default")
+	assert.Equal(t, snap.Resources[1].URN.Name(), "resA")
+	assert.Equal(t, snap.Resources[2].URN.Name(), "resB")
+
+	opts.UpdateOptions.Excludes = deploy.NewUrnTargetsFromUrns([]resource.URN{
+		"urn:pulumi:test::test::pulumi:providers:pkgA::default",
+	})
+
+	snap, err = lt.TestOp(Destroy).RunStep(project, p.GetTarget(t, snap), opts, false, p.BackendClient, nil, "2")
+
+	require.NoError(t, err)
+	require.Len(t, snap.Resources, 1)
+	assert.Equal(t, snap.Resources[0].URN.Name(), "default")
+}
+
+// We should be able to create a simple `A > B > C > D` hierarchy, destroy `B`,
+// and exclude both `C` as a dependent and `D` as a transitive dependent.
```

Comment:
I'm a bit confused by this description.  Not sure what it means to exclude C and D here?  We "exclude" A, but `C` and `D` are also destroyed, and thus not excluded?